### PR TITLE
[Feature] Stash unsubmitted changes after due date has passed

### PR DESCRIPTION
**Pull request by fde312**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context

We want to display only the changes that the student submitted (all changes that are on BitBucket). To ensure this, we will stash all unsubmitted changes.

### Description

- Implemented method to  stash changes after the due date has passed
- Invoke stash changes method in a scheduler, that will be executed when the due date has passed (same as for lock repo)

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Create Programming Exercise, set due date in the near future and enable manual assessment
4. Participate on exercise
4.1 Submit code and make change and just click on save (do not submit a second time!!!)
4.2 Just save your changes (do not submit at all)
4.3 Make changes and alway submit your code
5. Wait until the due date has passed
6. Check your submission in the code editor and check that the not saved changes are not displayed anymore

### Test Coverage

* GitService.java: 80%
* ProgrammingExerciseParticipationService,java: 86%

### Screenshots

During participation after clicking on save:
image

After due date:
image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java [hunk 1]
```diff
@@ -227,20 +227,25 @@ private Runnable lockStudentRepositories(ProgrammingExercise exercise, Predicate
...
                 if (programmingExercise.isEmpty()) {
                     throw new EntityNotFoundException("programming exercise not found with id " + programmingExerciseId);
                 }
-                if (numberOfFailedLockOperations > 0) {
+                if (numberOfFailedLockOperations > 0 || numberOfFailedStashOperations > 0) {
                     groupNotificationService.notifyInstructorGroupAboutExerciseUpdate(programmingExercise.get(),
-                            Constants.PROGRAMMING_EXERCISE_FAILED_LOCK_OPERATIONS_NOTIFICATION + failedLockOperations.size());
+                            Constants.PROGRAMMING_EXERCISE_FAILED_LOCK_OPERATIONS_NOTIFICATION + numberOfFailedLockOperations
+                                    + ". Furthermore, When stashing the not submitted changes, not all stash operations were successfull. Number of failed stash operations: "
+                                    + numberOfFailedStashOperations);
```
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java:244]: Reformulate / adjust reporting logic to notify only about operations that failed (not always both).

**Review by kloessst — COMMENTED** [review]
Already tested on this on TS1. Nice feature works really well.

**Comment by fde312** [comment]
I am not quite sure why the server tests on bamboo fail. I ran the test 100 times locally and everything works fine:
image

**Review by stefanwaldhauser — APPROVED** [review]
Works as expected :) I am not an export in the programming exercise code but also looks fine
image
image

**Comment by FrankeLukas** [comment]
Is there a specific reason we stash changes instead of tagging a release once the deadline is reached and always display that release? (Which has the advantage of not 'removing' things.) Might be worth considering for the long term?

**Review by kloessst — CHANGES_REQUESTED** [review]
Works fine. Cool feature. Just a small note about closing the repos:

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 2]
```diff
@@ -705,4 +705,16 @@ public boolean repositoryAlreadyExists(URL repoUrl) {
...
+    /**
+     * Stashes not submitted/committed changes of the repo.
+     *
+     * @param repo student repo of a participation in a programming exercise
+     * @throws GitAPIException
+     */
+    public void stashChanges(Repository repo) throws GitAPIException {
+        Git git = new Git(repo);
+        git.stashCreate().call();
+        git.close();
```
[the remark this answers is not in the document]
Comment (kloessst) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:718]: Nevermind the Git(repo, closeRepo) constructor is package private and can't be used. We would need to use Git.open(File dir, FS fs) the get a Git instance which closes the repo. You could still close the repo by hand if you don't need it anymore.

**Review by kloessst — APPROVED** [review]
Closing of repos shouldn't be an issue so I approve 😄

**Review by tiffca — APPROVED** [review]
Tested on test server 1 and confirmed only submitted changes are saved and actually graded. Saved changes are not available after the due date, only submitted changed. Tested the 3 scenarios described and a 4th scenario with not saving or submitting the changes.

**Review by derLalla — APPROVED** [review]
this is a nice feature! Approved :)

**Review by MaisiKoleni — APPROVED** [review]
Code still looks good as far as I can see.

**Comment by krusche** [comment]
> Is there a specific reason we stash changes instead of tagging a release once the deadline is reached and always display that release? (Which has the advantage of not 'removing' things.) Might be worth considering for the long term?

We wanted to have a quick solution. Stashing the changes means they would still be available for server admins and would not be lost. For students and instructors, they would currently not be visible.

One future extension could be to commit and push them into a different branch, however this would include multiple operations (which could potentially fail) and might have other side effects (e.g. a pushed branch could lead to a build on the CI server and to a result on Artemis which would have to be ignored, etc.)

**Review by krusche — APPROVED** [review]

**Pull request merged**